Refactor mobile navigation stack

brianlovin opened the pull request:
<!-- FILL OUT THE BELOW FORM OR YOUR PR WILL BE AUTOMATICALLY CLOSED -->
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- mobile

Damn, this one feels *really* good to figure out. 
1. Makes certain modals globally accessible, so that you can easily init a `ThreadComposer` or `DirectMessageComposer` from anywhere in the app. E.g. you might init a `ThreadComposer` from the home feed, but also from a community profile view.
2. Fixes the `tabBarVisible` logic to actually work! The tab bar is now hidden on thread and dm thread views.

This has brought up some new issues that I'll have to figure out, specifically:
1. Because you can now initiate a direct message from outside of the message tab, we are missing certain fields in the apollo cache that we always assumed would be there. Our assumption - and implementation of messages - on the web is that a user would *always* have to navigate to the messages tab in order to send a message. This would populate the `User` object cache with the `directMessageThreadsConnection` field - this field is what we update in our graphql mutations to append new direct message threads when a new one is composed.

We can no longer assume that this field exists in the cache, so we'll probably want to handle this by doing a background fetch of the users `directMessageThreadConnection` whenever the thread composer is opened.

2. We need to add some logic to post-publish flows in the modals to direct the user back to where they were. For example, in the direct messages example, here's what we have in our `onSubmit` method:

```js
onSubmit = (args) => {
  this.props
    .createDirectMessageThread(args)
    .then(result => {
      this.props.navigation.navigate('DirectMessageThread', {
        id: result.data.createDirectMessageThread.id,
      });
    })
};
```

The problem with this logic is that the following will happen:
1. You go to a user profile
2. You init a direct message - the dm composer modal opens
3. You sent the direct message
4. You are navigated to the `DirectMessageThread` view which is inside the `DirectMessages` stack navigator. All of a sudden you've been plucked from whatever context you were in within the app and now have been pushed forcefully into the Messages tab. Not good.

What we want to happen here is that if you init a DM from the Messages tab, it *should* push you to the conversation `onSubmit`. However, if you int a DM from anywhere else in the app, the `onSubmit` handler should tell the modal to close and dispatch a toast to say that the message was sent in the background. This way the user will retain their position in whatever navigation stack they're currently in without losing context.

I'll work on both of these problems now in this same PR, since they are fairly important. Writing them has been helpful to think the problem through :)

brianlovin committed:
Fix hiding tabbar on certain routes

brianlovin committed:
Rename function

brianlovin committed:
Refactor app navigation to support globally-accessible modals

brianlovin committed:
Rename tab bar icons file

brianlovin committed:
Clean up stack navigators to fit into new globally accessible modals nav pattern

brianlovin commented:
Quick mockup of the new setup:

<img width="1103" alt="screenshot 2018-06-21 14 08 57" src="https://user-images.githubusercontent.com/1923260/41745690-ab4ae8ba-755c-11e8-86f2-fbf2e78864f6.png">

Where each tab in the tab navigator extends our base stack, which contains all the navigation configurations for things like `Thread`, `Community`, `Channel` views, etc.

brianlovin committed:
Only show message author option when not current user

brianlovin committed:
Fix cache issue if initing a dm thread from outside of messages tab

brianlovin committed:
Allow DMs to be sent from anywhere without losing navigation context

brianlovin commented:
Okay those last few commits take care of all the issues above. Note @mxstbr on https://github.com/withspectrum/spectrum/pull/3379/commits/03367e7c33fb4dd751e34ea1b855add8831c347e#diff-da7e7f4d2b8d6502f9490112fc2362a8R25 => let me know what you think of using this pattern to determine the entry point. Otherwise there's not an easy way to programmatically access the previous route the user was viewing in the navigation stack, unless we were to do something crazy like sync navigation state with redux to keep a history. This way is *much* easier and gets accessed like this: https://github.com/withspectrum/spectrum/pull/3379/commits/03367e7c33fb4dd751e34ea1b855add8831c347e#diff-83552a414a3410ca633f685789a347e8R150

mxstbr reviewed (changes requested):
Great idea with the root modal navigator that renders the TabBar as the default route, very very neat solution! :100: :100: :100: 

Two comments:

> the directMessageThreadsConnection field - this field is what we update in our graphql mutations to append new direct message threads when a new one is composed.
> We can no longer assume that this field exists in the cache, so we'll probably want to handle this by doing a background fetch of the users directMessageThreadConnection whenever the thread composer is opened.

This is probably not a great solution, it introduces a race condition: what if the query doesn't complete until the user presses "Publish"? (slow network, short thread, whatever) App crash. We should solve this at a query level by creating the `user.directMessageThreadConnection` if it doesn't exist from the `createDirectMessageThread` mutation updater.

> However, if you int a DM from anywhere else in the app, the onSubmit handler should tell the modal to close and dispatch a toast to say that the message was sent in the background. This way the user will retain their position in whatever navigation stack they're currently in without losing context.

I see the thinking behind this decision, but honestly trying this it feels very unexpected to me—I think we should stick to the original behavior. (which also requires no extra code, as a bonus)

If I DM a person, no matter where I am in the app, I expect to see the DM thread afterwards. I want to be able to send follow-up messages, (e.g. personally, I'd usually send "Hey" and then "What's up?" as two separate messages) but also it's just what I _expect_ when I hit "Send" on that DM.

I vote that we stick to the original behavior there.

brianlovin commented:
> We should solve this at a query level by creating the user.directMessageThreadConnection if it doesn't exist from the createDirectMessageThread mutation updater.

I'll poke at this, I think I know what you mean.

> If I DM a person, no matter where I am in the app, I expect to see the DM thread afterwards. I want to be able to send follow-up messages, (e.g. personally, I'd usually send "Hey" and then "What's up?" as two separate messages) but also it's just what I expect when I hit "Send" on that DM.

Talked offline. The ideal behavior here is that if I dm a user from somewhere outside of the messages tab it will keep me in the conversation, but whenever I'm done it will keep me in my previous context. This will require having a modified dm composer that behaves very similar to Twitter - if I init a dm with someone through their profile, we'll do a query to see if a dm thread with that person exists before and if so, load all the messages for it so people can chat right there.

brianlovin committed:
Merge branch 'alpha' of github.com:withspectrum/spectrum into refactor-mobile-navigation-stack

brianlovin committed:
Add API method to look up a dm thread by user id

brianlovin committed:
Add graphql query to look up existing dm thread by userid

brianlovin committed:
Refactor dm composer to look up an existing thread if preloaded with a user id

brianlovin committed:
Cancel => close

brianlovin commented:
@mxstbr this is ready for another review - here's where I got:

1. Added a new query that takes a `userId` argument and returns a direct message thread, if one exists between the current user and that `userId`. Let me know how that feels to you.
2. I changed the DM composer behavior to the following:

a) If you go to the messages tab, press compose, and send a message you will go to that conversation. Pressing back goes to message stab.
b) If you init the composer from outside of the messages tab (e.g. press + hold on a thread in your feed and select the 'message author' option) it will do an initial lookup to see if you have an existing thread with that user. If you do, it will open the thread directly so you could chat, send multiple messages, etc. The cancel action will close the modal and keep you in whatever context you were in
c) If you init the composer from outside of the messages tab and you *don't* have an existing thread with that user, it will just open the composer. Sending a message will redirect to the thread view and the back action will take you to the messages tab (out of context).

IMO this is a halfway fix, but it's not worth spending a lot of time refactoring the rest of this. The reason this is a halfway fix is because I don't think we should have ever had a `createDirectMessageThread` mutation - instead, what would have been ideal is for us to simply let the `addDirectMessage` mutation create the thread on the server if no thread id exists in the mutation variables. This would allow step 2c above to be much cleaner: if we detect you don't have an existing thread with the user, it will just open an empty "fake" thread - sending the first message will create the dm thread in the backend and on the client side you can just keep typing multiple messages during the same step.

If you feel that's important, I'm happy to do that work - I just feel it's a rather chunky change for low benefits, when the state of this branch at least solves a lot of the most common flows somewhat well.

brianlovin committed:
Switch render logic to better fall back to composer

mxstbr commented on api/queries/directMessageThread/rootDirectMessageThreadByUserId.js:
```diff
@@ -0,0 +1,22 @@
+  const { user: currentUser, loaders } = ctx;
+  const { userId } = args;
+
+  if (!currentUser || !currentUser.id) return null;
```
Don't we have a `checkAuth` wrapper for resolvers for this?

mxstbr reviewed (commented):
Testing now!

mxstbr commented on mobile/views/DirectMessageComposer/Composer.js:
```diff
@@ -0,0 +1,214 @@
+            returnKeyType="search"
+            autoFocus
+            innerRef={elem => (this.searchInput = elem)}
+            placeholder={'Search for people...'}
```
nit: no autocorrect and autocapitalize

brianlovin replied:
Do we want those on user search? Autocorrect may be more annoying than not when searching for names

mxstbr replied:
That's why I said _no_ autocorrect and autocapitalize—right now it autocorrects for me.

mxstbr commented on mobile/views/DirectMessageThread/index.js:
```diff
@@ -1,39 +1,141 @@
 
-class DirectMessageThreadView extends React.Component<Props> {
+class DirectMessageThread extends Component<Props> {
+  messagesComponent: any;
```
nit: this isn't a thing, right?

mxstbr commented on api/types/DirectMessageThread.js:
```diff
@@ -31,6 +31,7 @@ const DirectMessageThread = /* GraphQL */ `
 
 	extend type Query {
 		directMessageThread(id: ID!): DirectMessageThread
+		directMessageThreadByUserId(userId: ID!): DirectMessageThread
```
Does this need a separate query? Imo querying with an object that has either an `id` field or an `userId` field makes a lot more sense, if that's possible. Maybe something like this works?

```GraphQL
extend type Query {
  type DirectMessageThreadQueryInput {
    id: ID
    userId: ID
  }
  directMessageThread(input: ID! | DirectMessageThreadQueryInput): DirectMessageThread
}
```

brianlovin replied:
While that *is* a possible way to accomplish this, I'm honestly not sure if it's better. I feel that a common mistake we've made in our queries and mutations is to try and be too clever by bundling multiple possible intentions inside a single action. For example we have `toggleChannelSubscription` instead of two: `subscribeToChannel` and `unsubscribeFromChannel` or something like that - as a result, our mutation is much much more complex.

In the same way, I worry that trying to do two things at once here may be confusing.

>  instead, what would have been ideal is for us to simply let the addDirectMessage mutation create the thread on the server if no thread id exists in the mutation variables.

^^ I said this above, but feel less good about that idea now with this in mind. It would require over-complicating the input variables that the mutation requires. I'd much prefer to have a strong set of single-purpose queries and mutations that are easy to read and refactor, versus trying to be clever and combine multiple use cases in one.

What do you think?

brianlovin replied:
Talked offline and agreed :)

mxstbr reviewed (commented):
This is super nice, well done! :clap: :clap: 

Just some small nitpicks inline

brianlovin committed:
Use requireauth wrapper

brianlovin committed:
Remove unnused class property typing

brianlovin committed:
Merge branch 'alpha' of github.com:withspectrum/spectrum into refactor-mobile-navigation-stack

brianlovin committed:
Dont autocorrect names
